diff --git a/CMakeLists.txt b/CMakeLists.txt
index 1a0e60a2..5d145863 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -149,9 +149,7 @@ if (WIN32)
         "Verify signature of a dependency during its dynamic linking at run-time on Windows*" OFF)
 endif()
 
-if (NOT DEFINED BUILD_SHARED_LIBS)
-    set(BUILD_SHARED_LIBS ON)
-endif()
+set(BUILD_SHARED_LIBS ON)
 
 if (NOT BUILD_SHARED_LIBS)
     if(NOT DEFINED CMAKE_POSITION_INDEPENDENT_CODE)
